apps/condo/domains/resident/tasks/helpers/sendBillingReceiptsAddedNotifications.spec.js

```diff
@@ -3,178 +3,226 @@
+            expect(messages2).toHaveLength(1)
+            expect(messages1[0].organization.id).toEqual(resident1.organization.id)
+            expect(messages2[0].organization.id).toEqual(resident2.organization.id)
+        })
+
+        test('Should not create duplicate notifications if task is executed multiple times within the same period', async () => {
+            const { receipt, resident, billingContext } = await makeBillingReceiptWithResident({ toPay: '15000' })
+
+            const date = dayjs(receipt.createdAt).subtract(1, 'hour').toISOString()
+            await sendBillingReceiptsAddedNotificationForOrganizationContext({ ...billingContext, organization: billingContext.organization.id }, date)
+            await sendBillingReceiptsAddedNotificationForOrganizationContext({ ...billingContext, organization: billingContext.organization.id }, date)
 
-            const notificationKey = makeMessageKey(receipt.period, receipt.account.number, receipt.category.id, resident.id)
             const messageWhere = {
+                user: { id: resident.user.id },
                 type: BILLING_RECEIPT_ADDED_TYPE,
-                uniqKey: notificationKey,
             }
-            const message = await Message.getOne(admin, messageWhere)
 
-            expect(message).toBeUndefined()
+            const messages = await Message.getAll(admin, messageWhere)
+            expect(messages).toHaveLength(1)
+            expect(messages[0].organization.id).toEqual(resident.organization.id)
         })
 
+        test('Should not send duplicate notifications from different contexts', async () => {
+            const environment = new TestUtils([ResidentTestMixin])
+            await environment.init()
+            const { receipt, resident, billingContext, residentUser } = await makeBillingReceiptWithResident({ toPay: '5000' }, false)
+            const { receipt: receipt1, resident: resident1, billingContext: billingContext1, residentUser: residentUser1 } = await makeBillingReceiptWithResident({ toPay: '5000' }, false, residentUser)
+
+            await registerBillingReceiptsByTestClient(admin, {
+                context: { id: billingContext.id },
+                receipts: environment.createJSONReceipt(),
+            })
+
+            await registerBillingReceiptsByTestClient(admin, {
+                context: { id: billingContext1.id },
+                receipts: environment.createJSONReceipt(),
+            })
+
+            expect(resident.user.id).toEqual(resident1.user.id)
+            
+            await sendBillingReceiptsAddedNotificationForOrganizationContext({ ...billingContext, organization: billingContext.organization.id }, dayjs(receipt.createdAt).subtract(1, 'hour').toISOString())
+            await sendBillingReceiptsAddedNotificationForOrganizationContext({ ...billingContext1, organization: billingContext1.organization.id }, dayjs(receipt1.createdAt).subtract(1, 'hour').toISOString())
```

Comment:
What if the second task starts until the first one is finished?
Could you please add a test if such situation is possible.

Reply:
Execution order is not important. I need send a push from any task but only one time